Init campaigns directory

lewri opened the pull request:
Segment of #2209 (aka the easy bit)

<!-- If your PR is still being drafted you must either submit it as a "Draft Pull Request" or add [WIP] to your title. -->

**Describe what the proposed change does**
- Initialises the campaigns directory on launch and adds it as a config line.
- Encompasses part of the prepwork for CommunityContent.

lewri commented:
This line probably should've also come across:
https://github.com/CorsixTH/CorsixTH/blob/1ee471158613ed40537e49445907fd67d107412d/CorsixTH/Lua/app.lua#L648

Alberth289346 commented on CorsixTH/Lua/app.lua:
```diff
@@ -396,6 +398,28 @@ function App:initUserLevelDir()
+    self.user_campaign_dir = self.user_campaign_dir .. pathsep
+  end
+  return true
+end
```
Don't copy/paste code!!

Refactor `App:initUserLevelDir` to a common part and a userLevel-specific part, and use the former also here.

lewri replied:
Tbh yes that was me trying to just get it working ages ago. Then neglected to actually tidy it. I'll work on that.

Alberth289346 commented on CorsixTH/Lua/app.lua:
```diff
@@ -616,9 +640,8 @@ end
-  local path = debug.getinfo(1, "S").source:sub(2, -12)
   -- First look in Campaigns. If not found there, fall back to Levels.
-  local list_of_possible_paths = { self.user_level_dir, path .. "Campaigns", self.level_dir }
+  local list_of_possible_paths = { self.user_campaign_dir, self.campaign_dir, self.user_level_dir, self.level_dir }
```
Not your fault, but yuck.

"list_of" is type information that is already expressed at the right of the `=` sign. Also it's clear the right side that it's a list of paths, the name doesn't need to tell that too. The reader now has to check what happens with the variable to understand why it exists.

Instead, use it to express the reason of having the variable. Something like "paths_to_search" says what the intention of that data is. From that intention it's fully logical that you iterate over the values next.

Alberth289346 commented on CorsixTH/Lua/app.lua:
```diff
@@ -118,7 +118,9 @@ function App:init()
   local good_install_folder, error_message = self:checkInstallFolder()
   self.good_install_folder = good_install_folder
   self.level_dir = debug.getinfo(1, "S").source:sub(2, -12) .. "Levels" .. pathsep
+  self.campaign_dir = debug.getinfo(1, "S").source:sub(2, -12) .. "Campaigns" .. pathsep
```
Not your fault, but `debug.getinfo(1, "S").source:sub(2, -12)`?

What is debugging code doing here? Why isn't there a nice function-name for this madness?

lewri replied:
Unfortunately I can't answer. And I would assume as I can't I probably shouldn't try to guess?

lewri replied:
@tobylane answered in Discord. Toby I'm happy if you wanted to commit that change into this PR if you wanted?

tobylane replied:
https://grep.app/search?q=Debug.getinfo&filter[repo][0]=CorsixTH/CorsixTH&filter[lang][0]=Lua&filter[path][0]=CorsixTH/

There's 19 uses of it. It can go here, but if you're ready before then it can go in a new PR.

lewri replied:
There's one in `CorsixTH.lua` -- not much you can do about that one in app, unless you set it in that file.

lewri replied:
If you're touching the entire codebase, I'd put it in a new PR

Alberth289346 commented:
Sorry, much comment is not about your changes directly, but cleaning up code that you touch is always very useful.

lewri commented:
> Sorry, much comment is not about your changes directly, but cleaning up code that you touch is always very useful.

Fixed two of these, but not the `debug` one.

I added a TODO that I will put into a new PR once this one merges... just in case I forget to do it.

lewri commented on CorsixTH/Lua/app.lua:
```diff
@@ -374,26 +375,38 @@ function App:init()
-  end
-  return true
+
+  self.user_level_dir = self.config.levels or
+      conf_path:match("^(.-)[^" .. pathsep .. "]*$") .. "Levels"
+  self.user_level_dir = setUserDir(self.user_level_dir, "Levels")
+  self.user_campaign_dir = self.config.campaigns or
+      conf_path:match("^(.-)[^" .. pathsep .. "]*$") .. "Campaigns"
+  self.user_campaign_dir = setUserDir(self.user_campaign_dir, "Campaigns")
```
Will condense more in a future PR

Alberth289346 replied:
Insert an empty line below 406 so it's easier to understand it's the same thing twice but with a different kind of directly.

Re-use of `self.usr_xyz_dir` variable is confusing, change the first assignment to a local variable.

If you stick the common parts in a function and supply the non-common parts as parameters, future PR would be pretty much done?

lewri replied:
That was the plan for the future PR, yes -- mainly to go with https://github.com/CorsixTH/CorsixTH/pull/2273#issuecomment-1287763941

Alberth289346 replied:
ok, fair enough.

lewri commented on CorsixTH/Lua/app.lua:
```diff
@@ -374,26 +375,38 @@ function App:init()
-  if self.user_level_dir:sub(-1, -1) ~= pathsep then
-    self.user_level_dir = self.user_level_dir .. pathsep
-  end
-  return true
```
Should the `return true` had been left in?

Alberth289346 commented on CorsixTH/Lua/app.lua:
```diff
@@ -374,25 +375,39 @@ function App:init()
+    if lfs.attributes(dir, "mode") ~= "directory" then
+      if not lfs.mkdir(dir) then
+        print("Notice:" .. label .. "directory does not exist and could not be created.")
+        return false -- Force game exit
```
This is not what you promised at line 387.

Alberth289346 commented on CorsixTH/Lua/app.lua:
```diff
@@ -616,10 +631,9 @@ end
   -- First look in Campaigns. If not found there, fall back to Levels.
-  local list_of_possible_paths = { self.user_level_dir, path .. "Campaigns", self.level_dir }
-  for _, parent_path in ipairs(list_of_possible_paths) do
+  local paths_to_search = { self.user_campaign_dir, self.campaign_dir, self.user_level_dir, self.level_dir }
```
```suggestion
  local paths_to_search = {
    self.user_campaign_dir,
    self.campaign_dir,
    self.user_level_dir,
    self.level_dir
  }
```
Maybe too many lines?

Alberth289346 replied:
I considered that for about 2 seconds too, but it's not our code style.

Also, if you don't see it when you have 1 word at each line, would you see it if you have multiple words at each line?
I think the benefits are doubtful, and not enough to break code style.

Alberth289346 commented on CorsixTH/Lua/config_finder.lua:
```diff
@@ -414,12 +414,13 @@ unicode_font = nil -- [[X:\ThemeHospital\font.ttc]]
--- be used for storing new maps / levels in. Should this not be suitable, then
--- uncomment the following line, and point it to a directory which exists and
--- is more suitable.
+-- Levels and Campaigns. By default, the "Levels" and "Campaigns" directory alongside
```
```suggestion
-- Levels and Campaigns. By default, the "Levels" and "Campaigns" directory next to
```
?

Alberth289346 commented on CorsixTH/Lua/config_finder.lua:
```diff
@@ -414,12 +414,13 @@ unicode_font = nil -- [[X:\ThemeHospital\font.ttc]]
--- uncomment the following line, and point it to a directory which exists and
--- is more suitable.
+-- Levels and Campaigns. By default, the "Levels" and "Campaigns" directory alongside
+-- this config file will be used for storing new maps / levels / campaigns in. Should
+-- this not be suitable, then uncomment the following lines, and point it to a directory
```
```suggestion
-- this config file will be used for storing new maps / levels / campaigns. If this is
-- not be suitable, then uncomment the following lines, and point it to a directory.
```
I am a little lost where the maps are going, as there are two directories and three things to store.

lewri replied:
I don't think I paid enough attention to this textual comment. It needs to say that created maps go into the Levels folder currently.

Alberth289346 commented:
LGTM

TheCycoONE commented on CorsixTH/Lua/app.lua:
```diff
@@ -616,10 +632,14 @@ end
+  local paths_to_search = {
+    self.user_campaign_dir,
+    self.campaign_dir,
+    self.user_level_dir,
```
The comment is misleading, the user_level_dir should be between the user_campaign_dir and the campaign_dir

lewri replied:
Order changed :+1:

lewri replied:
Though our comment should be something like: First check user's campaigns/level directories. If not found there, fall back to default directories.

lewri replied:
Comment likely isn't needed at all now, will remove.

lewri commented:
@TheCycoONE you may need to fix the commit history of this one before merge, Github's website didn't do a proper rebase by the look of it.
